packages/runtime-vapor/src/componentSlots.ts

```diff
@@ -0,0 +1,86 @@
+import { type IfAny, isArray } from '@vue/shared'
+import { baseWatch } from '@vue/reactivity'
+import type { ComponentInternalInstance } from './component'
+import type { Block } from './apiRender'
+import { createVaporPreScheduler } from './scheduler'
+
+// TODO: SSR
+
+export type Slot<T extends any = any> = (
+  ...args: IfAny<T, any[], [T] | (T extends undefined ? [] : never)>
+) => Block
+
+export type InternalSlots = {
+  [name: string]: Slot | undefined
+}
+
+export type Slots = Readonly<InternalSlots>
+
+export interface DynamicSlot {
+  name: string
+  fn: Slot
+  key?: string
+}
+
+export type DinamicSlots = () => (DynamicSlot | DynamicSlot[])[]
+
+export const initSlots = (
+  instance: ComponentInternalInstance,
+  rawSlots: InternalSlots | null = null,
+  dynamicSlots: DinamicSlots | null = null,
+) => {
+  const slots: InternalSlots = {}
+  for (const key in rawSlots) {
+    Object.defineProperty(slots, key, {
+      get: () => rawSlots[key],
+      enumerable: true,
+    })
+  }
```

Comment:
```suggestion
  const slots: InternalSlots = extend({}, rawSlots)
```

There's no side effect when reading property of `rawSlots`.